mozilla-services/syncstorage-rs
# refactor: Tokenserver: Add mature MySQL adapter

## Description

Adds mature infrastructure for a Tokenserver-specific MySQL database.

## How should reviewers test?
* Create an account on FxA stage: https://accounts.stage.mozaws.net/signup
* Set `identity.fxaccounts.remote.oauth.uri` to https://oauth.stage.mozaws.net
* Set `identity.fxaccounts.remote.root` to https://accounts.stage.mozaws.net
* Set `identity.fxaccounts.auth.uri` to https://api-accounts.stage.mozaws.net/v1
* Set `identity.fxaccounts.remote.profile.uri` to https://profile.stage.mozaws.net/v1
* Set `identity.sync.tokenserver.uri` to http://localhost:8000/1.0/sync/1.5
* Create a Tokenserver database:
```sql
CREATE DATABASE tokenserver_rs;
```
* Add a service and a node to your database:
```sql
INSERT INTO services (service) VALUES("sync-1.5");
INSERT INTO nodes (service, node, available, capacity) VALUES([service ID from previous step], "http://localhost:8000/", 100, 100);
```
* Set the following values in local.toml:
```toml
enable_tokenserver = true
tokenserver.database_url = "mysql://sample_user:sample_password@localhost/tokenserver_rs"
tokenserver.fxa_oauth_server_url = "https://oauth.stage.mozaws.net"
```
* Add `println!("{}", user_email)` to L47 in `src/tokenserver/handlers.rs`
* Start syncstorage via `make run`
* Restart Firefox, log in to sync, and attempt a sync
* Using the email address from the server logs, add a user to the database:
```sql
INSERT INTO users (service, email, generation, client_state, created_at, nodeid) VALUES([service ID from prior step], "[email address from logs]", 0, "yes", 1620924095, [node ID from prior step]);
```
* Remove the `println!` and recompile/restart the server
* Attempt a sync
* Ensure that the sync was successful via about:sync-log

## Issue(s)

Closes #1054
